### DISCUSSION TITLE
RAUC kernel update failure (Error:FAT-fs bogus number of FAT structure)

### DISCUSSION THREAD

**Discussion by Rajalakshmi-Ranganathan** [rauc/meta-rauc#311] [Q&A]
Hi @ejoerns and others,

I'm using meta-rauc (Honister) where I am able to successfully update the rootfs slots using rauc bundle. But I am facing below issue when trying to update kernel. Also I have provided details like partition table, system.conf & bundle recipe file. Please take a look and provide your suggestions.

### Error:

rauc --debug install http://10.42.0.2:8080/kernel_update/update-bundle-zynq-generic.raucb
rauc-Message: 12:37:50.434: Debug log domains: 'rauc'
(rauc:872): rauc-DEBUG: 12:37:50.454: install started
(rauc:872): rauc-DEBUG: 12:37:50.455: input bundle: http://10.42.0.2:8080/kernel_update/update-bundle-zynq-generic.raucb
(rauc:872): rauc-DEBUG: 12:37:50.503: Trying to contact rauc service
installing
  0% Installing
  0% Determining slot states
 20% Determining slot states done.
 20% Checking bundle
 20% Verifying signature
 40% Verifying signature done.
 40% Checking bundle done.
 40% Checking manifest contents
 60% Checking manifest contents done.
 60% Determining target install group
 80% Determining target install group done.
 80% Updating slots
 80% Checking slot rootfs.1
 85% Checking slot rootfs.1 done.
 85% Copying image to rootfs.1
 90% Copying image to rootfs.1 done.
 90% Checking slot kernel.1
 95% Checking slot kernel.1 done.
 95% Copying image to kernel.1
 99% Copying image to kernel.1 done.
**FAT-fs (mmcblk0p6): bogus number of FAT structure
 99% Updating slots failed.
100% Installing failed.
LastError: Installation error: Error while writing status file: failed to mount slot: failed to run mount: Child process exited with code 32
Installing `http://10.42.0.2:8080/kernel_update/update-bundle-zynq-generic.raucb` failed**

### My Disk Partition Table:
**Device         Boot    Start      End  Sectors  Size Id Type**
/dev/mmcblk0p1          2048   206847   204800  100M  c W95 FAT32 (LBA)   ======> BOOT
/dev/mmcblk0p2        206848  2303999  2097152    1G  c W95 FAT32 (LBA)  ======> BOOTENV
/dev/mmcblk0p3       2304000 30930943 28626944 13.7G  5 Extended
/dev/mmcblk0p5       2306048  2510847   204800  100M  c W95 FAT32 (LBA) =====> Kernel.0 Image
/dev/mmcblk0p6       2512896  2717695   204800  100M  c W95 FAT32 (LBA) =====> Kernel.1 Image
/dev/mmcblk0p7       2719744 15302655 12582912    6G 83 Linux                  =====> Rootfs.0 files
/dev/mmcblk0p8      15304704 30930943 15626240  7.5G 83 Linux                =====> Rootfs.1 files

### RAUC system.conf:

 [system]
 compatible=RAUC Fw Update
 bootloader=uboot
 bundle-formats=verity
 max-bundle-download-size=209715200

 [keyring]
 path=/etc/rauc/ca.cert.pem

 [slot.rootfs.0]
 device=/dev/mmcblk0p7
 type=ext4
 bootname=system0

 [slot.rootfs.1]
 device=/dev/mmcblk0p8
 type=ext4
 bootname=system1

 [slot.kernel.0]
 device=/dev/mmcblk0p5
 type=vfat
 parent=rootfs.0

 [slot.kernel.1]
 device=/dev/mmcblk0p6
 type=vfat
 parent=rootfs.1

### My Bundle Recipe:

DESCRIPTION = "RAUC bundle generator"

LICENSE = "MIT"
LIC_FILES_CHKSUM = "file://${COMMON_LICENSE_DIR}/MIT;md5=0835ade698e0bcf8506ecda2f7b4f302"

inherit bundle

RAUC_BUNDLE_FORMAT = "verity"

RAUC_BUNDLE_COMPATIBLE = "RAUC Fw Update"
RAUC_BUNDLE_VERSION = "v20240326"
RAUC_BUNDLE_DESCRIPTION = "RAUC Update Bundle"

RAUC_BUNDLE_SLOTS = "rootfs kernel"
RAUC_SLOT_rootfs = "petalinux-image-minimal"
RAUC_SLOT_rootfs[fstype] = "tar.gz"

RAUC_SLOT_kernel = "linux-xlnx"
RAUC_SLOT_kernel[type] = "file"
RAUC_SLOT_kernel[file] = "fitImage"

RAUC_KEY_FILE = "${THISDIR}/files/development-1.key.pem"
RAUC_CERT_FILE = "${THISDIR}/files/development-1.cert.pem"

**Comment by jluebbe** [comment]
You've configured the the `kernel.*` slots as type `vfat` and have not configured a [shared data directory](https://rauc.readthedocs.io/en/latest/reference.html#data-directory), which means that RAUC will try to mount them to store slot status information. But, because you write the FIT image directory to that slot, it doesn't contain a FAT file system and mounting fails.

I'd suggest to both:

- configure a data directory on a shared partition
- use type `raw` for the kernel slots

**Reply by Rajalakshmi-Ranganathan** [reply]
Thanks a lot @jluebbe for the response.

I'm using Rauc version 1.7 where **data-directory** key is not available. However I used **statusfile** key and able to see bundle installation completed successfully.

But once I reboot, kernel is unable to boot. After debugging further, I noticed that kernel.1 slot partition mmcblk0p6 got corrupted after "rauc update". Please find the log, system.conf and mount output as below and kindly help to resolve the partition corruption issue.

### Rauc update result log:

:~$ rauc --debug install http://10.42.0.2:8080/kernel_update/update-bundle-zynq-generic.raucb
rauc-Message: 12:51:05.871: Debug log domains: 'all'
(rauc:989): rauc-DEBUG: 12:51:05.881: install started
(rauc:989): rauc-DEBUG: 12:51:05.882: input bundle: http://10.42.0.2:8080/kernel_update/update-bundle-zynq-generic.raucb
(rauc:989): rauc-DEBUG: 12:51:05.899: Trying to contact rauc service
installing
  0% Installing
  0% Determining slot states
 20% Determining slot states done.
 20% Checking bundle
 20% Verifying signature
 40% Verifying signature done.
 40% Checking bundle done.
 40% Checking manifest contents
 60% Checking manifest contents done.
 60% Determining target install group
 80% Determining target install group done.
 80% Updating slots
 80% Checking slot rootfs.1
 85% Checking slot rootfs.1 done.
 85% Copying image to rootfs.1
 90% Copying image to rootfs.1 done.
 90% Checking slot kernel.1
 95% Checking slot kernel.1 done.
 95% Copying image to kernel.1
 99% Copying image to kernel.1 done.
 99% Updating slots done.
100% Installing done.
Installing `http://10.42.0.2:8080/kernel_update/update-bundle-zynq-generic.raucb` succeeded

### system.conf:

 [system]
 compatible=RAUC Fw Update
 bootloader=uboot
 bundle-formats=verity
 max-bundle-download-size=209715200
 **statusfile=/media/central-status.raucs**

 [keyring]
 path=/etc/rauc/ca.cert.pem

 [slot.rootfs.0]
 device=/dev/mmcblk0p7
 type=ext4
 bootname=system0

 [slot.rootfs.1]
 device=/dev/mmcblk0p8
 type=ext4
 bootname=system1

 [slot.kernel.0]
 device=/dev/mmcblk0p5
 **type=raw**
 parent=rootfs.0

 [slot.kernel.1]
 device=/dev/mmcblk0p6
 **type=raw**
 parent=rootfs.1


### Mount Error:
sudo mount /dev/mmcblk0p6 /tmp/kernel/
mount: /var/volatile/tmp/kernel: wrong fs type, bad option, bad superblock on /dev/mmcblk0p6, missing codepage or helper program, or other error.

**Reply by jluebbe** [reply]
Why do you want to mount /dev/mmcblk0p6, which contains a fit image (according to your bundle recipe, and not a filesystem? That won't work.

**Reply by Rajalakshmi-Ranganathan** [reply]
Hi @jluebbe 

In my project, the fitImage file name is "image.ub". My u-boot scripts and it's commands will load it only if it finds the file image.ub in the kernel vfat partition. So I am planning to use "rename" property as below to achieve it.

RAUC_SLOT_kernel[file] = "fitImage"
RAUC_SLOT_kernel[rename] = "image.ub"

So I'm trying to achieve kernel slot updates as same as rootfs.

During Rootfs slot update in Rauc:
1. Format and create ext4 partition
2. Mount the rootfs partition
3. Extract contents to the mounted folder
4. Unmount the partition

Current kernel slot update in Rauc:
1. Opening the partition slot
2. Writing the data

I think the kernel slot update actually does binary writing of fitImage contents but I wanted it to be a "cp" operation by mount and unmount a valid vfat filesystem.

Kindly help on this.

**Reply by florentsfl** [reply]
@jluebbe I understand that in this case the fitImage is written as is on the partition without a filesystem, but in this case, is the bootloader supposed to read the partition without filesystem to boot the fitImage ?

The question I have is how to boot the fitImage from the bootloader when it is written on a partition without filesystem, is it even possible ?

Because if we follow the instructions from the bundle.bbclass, it seems like to update the kernel slot we could use this https://github.com/rauc/meta-rauc/blob/8992a51e245b82fe59ade3da1bfac4ffde967d7b/classes-recipe/bundle.bbclass#L22 but then that would copy a raw kernel image to a partition so I wonder if it is bootable then.

Thank you for your help
